### PR TITLE
Fixed #28944 -- Allowed values()/values_list() after QuerySet.select_for_update(of=())

### PR DESCRIPTION
**Pull request by ChetanKhanna**
Refactored the PR that Anssi started. Improved the error message
and test. Thanks Anssi for the approach.

### CODE REVIEW

**Comment by blueyed** [comment]
Just for reference: https://code.djangoproject.com/ticket/28944

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -487,9 +487,10 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
         refcounts_before = self.query.alias_refcount.copy()
         try:
             extra_select, order_by, group_by = self.pre_sql_setup()
-            for_update_part = None
             # Is a LIMIT/OFFSET clause needed?
             with_limit_offset = with_limits and (self.query.high_mark is not None or self.query.low_mark)
+            # Calculate the update part - it might alter the joins
+            for_update_part = self.get_for_update_part(with_limit_offset)
```
Comment (felixxm) [inline django/db/models/sql/compiler.py]: It's not used only for combined queries so we should call it only in `not combinator` branch.

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -628,6 +602,36 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
             # Finally do cleanup - get rid of the joins we created above.
             self.query.reset_refcounts(refcounts_before)
 
+    def get_for_update_part(self, with_limit_offset):
```
Comment (felixxm) [inline django/db/models/sql/compiler.py]: `get_for_update_part()` -> `get_for_update_sql()`. We could move this hook in to a separate commit/PR.

Comment (ChetanKhanna) [reply django/db/models/sql/compiler.py]: Changed.

File: django/db/models/sql/compiler.py (hunk 1 above)
Comment (ChetanKhanna) [reply django/db/models/sql/compiler.py]: Sorry I couldn't catch that. Are you saying we should move `for_update_part` inside the `else` clause? But that's not possible since it is also used outside the else-block scope.

**Comment by ChetanKhanna** [comment]
Hi @felixxm
Not sure if its a good place to ask, but locally `test_async_process_template_response_returns_none_with_sync_client` failed with Traceback:

```
File "/tests/django/django/test/utils.py", line 381, in inner
    return func(*args, **kwargs)
  File "/tests/django/tests/middleware_exceptions/tests.py", line 267, in test_async_process_template_response_returns_none_with_sync_client
    self.client.get('/middleware_exceptions/template_response/')
  File "/tests/django/django/test/client.py", line 732, in get
    response = super().get(path, data=data, secure=secure, **extra)
  File "/tests/django/django/test/client.py", line 395, in get
    **extra,
  File "/tests/django/django/test/client.py", line 470, in generic
    return self.request(**r)
  File "/tests/django/django/test/client.py", line 709, in request
    self.check_exception(response)
  File "/tests/django/django/test/client.py", line 571, in check_exception
    raise exc_value
  File "/tests/django/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/tests/django/django/core/handlers/base.py", line 196, in _get_response
    middleware_method.__self__.__class__.__name__,
AttributeError: 'AsyncToSync' object has no attribute '__self__'
```

Is this a known issue? Or am I missing a module? I was testing via `django-docker-box`.

File: django/db/models/sql/compiler.py (hunk 1 above)
Comment (felixxm) [reply django/db/models/sql/compiler.py]: We should leave here `for_update_part = None` (in the same line as it was previously) and calculate it only in `not combinator` branch, because it's not used for combined queries.

**Comment by felixxm** [comment]
@ChetanKhanna Yes, it's not a good place to ask about unrelated test failure. Install `asgiref >= 3.2.5 ` (see https://github.com/django/asgiref/commit/43740dca083048f97c7aee43bd4b877eb3903dc3).

File: django/db/models/sql/compiler.py (hunk 1 above)
Comment (ChetanKhanna) [reply django/db/models/sql/compiler.py]: Done, thanks.

**Comment by ChetanKhanna** [comment]
Hi @felixxm! I rebased this from master, is anything else needed on this?

**Comment by felixxm** [comment]
@ChetanKhanna Thanks :+1: I rebased from master and moved `SQLCompiler.get_for_update_sql()` to a separate commit (now it's easier to review).
<summary> We have two failures on Oracle:</summary>
<details>

```
======================================================================
ERROR: test_for_update_sql_model_inheritance_generated_of (select_for_update.tests.SelectForUpdateTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/django/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/django/django/db/backends/oracle/base.py", line 530, in execute
    return self.cursor.execute(query, self._param_generator(params)) 
cx_Oracle.DatabaseError: ORA-00904: "SELECT_FOR_UPDATE_EUCOUNTRY"."ENTITY_PTR_ID": invalid identifier

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/django/django/test/testcases.py", line 1274, in skip_wrapper
    return test_func(*args, **kwargs)
  File "/django/tests/select_for_update/tests.py", line 147, in test_for_update_sql_model_inheritance_generated_of
    list(EUCountry.objects.select_for_update(of=('self',)))
  File "/django/django/db/models/query.py", line 287, in __iter__ 
    self._fetch_all()
  File "/django/django/db/models/query.py", line 1305, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/django/django/db/models/query.py", line 53, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/django/django/db/models/sql/compiler.py", line 1165, in execute_sql
    cursor.execute(sql, params)
  File "/django/django/db/backends/utils.py", line 98, in execute
    return super().execute(sql, params)
  File "/django/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/django/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/django/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/django/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/django/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/django/django/db/backends/oracle/base.py", line 530, in execute
    return self.cursor.execute(query, self._param_generator(params)) 
django.db.utils.DatabaseError: ORA-00904: "SELECT_FOR_UPDATE_EUCOUNTRY"."ENTITY_PTR_ID": invalid identifier

======================================================================
ERROR: test_for_update_sql_model_inheritance_ptr_generated_of (select_for_update.tests.SelectForUpdateTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/django/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/django/django/db/backends/oracle/base.py", line 530, in execute
    return self.cursor.execute(query, self._param_generator(params)) 
cx_Oracle.DatabaseError: ORA-00904: "SELECT_FOR_UPDATE_EUCOUNTRY"."ENTITY_PTR_ID": invalid identifier

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/django/django/test/testcases.py", line 1274, in skip_wrapper
    return test_func(*args, **kwargs)
  File "/django/tests/select_for_update/tests.py", line 158, in test_for_update_sql_model_inheritance_ptr_generated_of
    list(EUCountry.objects.select_for_update(of=('self', 'country_ptr',)))
  File "/django/django/db/models/query.py", line 287, in __iter__ 
    self._fetch_all()
  File "/django/django/db/models/query.py", line 1305, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/django/django/db/models/query.py", line 53, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/django/django/db/models/sql/compiler.py", line 1165, in execute_sql
    cursor.execute(sql, params)
  File "/django/django/db/backends/utils.py", line 98, in execute
    return super().execute(sql, params)
  File "/django/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/django/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/django/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/django/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/django/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/django/django/db/backends/oracle/base.py", line 530, in execute
    return self.cursor.execute(query, self._param_generator(params)) 
django.db.utils.DatabaseError: ORA-00904: "SELECT_FOR_UPDATE_EUCOUNTRY"."ENTITY_PTR_ID": invalid identifier
```
</details>

**Comment by felixxm** [comment]
Closing due to inactivity.
